### PR TITLE
[wasm] Typed Array support for better memory usage

### PR DESCRIPTION
**Pull request by kjpou1**
### Breaking changes
  - Remove NewArrayJS which is replaced with Array bidning.  call `new WebAssembly.Core.Array` constructor.
  - The Initialize bindings call is no longer needed.

Add Typed Array support to bindings.  This was prompted by the need to help with memory usage and transfer of typed array data more efficiently.  Typed Array memory can not be directly accessed but instead needs to be transferred to and from emscripten's heap.

- Refactored code to separate Core objects from JSObject.
- Add interfaces.
- Core Object support

NameSpace `WebAssembly.Core`

| Core Object |  Description |
| ------------- | -------------- |
| Array | Binds [Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects#Indexed_collections) |
| Int8Array | Binds [Int8Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects#Indexed_collections) | 
| Uint8Array| Binds [Uint8Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects#Indexed_collections) | 
| Uint8ClampedArray | Binds [Uint8ClampedArray ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects#Indexed_collections) | 
| Int16Array | Binds [Int16Array ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects#Indexed_collections) | 
| Uint16Array | Binds [Uint16Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects#Indexed_collections) | 
| Int32Array | Binds [Int32Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects#Indexed_collections) | 
| Uint32Array | Binds [Uint32Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects#Indexed_collections) | 
| Float32Array | Binds [Float32Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects#Indexed_collections) | 
| Float64Array | Binds [Float64Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects#Indexed_collections) | 
           
Helper methods have been added as extensions to these new objects.  These methods will allow the objects to work with other managed objects:  primitive arrays, ArraySegment<T> and Span<T>.

| Method | Description |
| --- | --- |
| CopyFrom | Copies from heap memory to typed array memory |
| CopyTo | Copies from typed array memory to heap memory |

### Build/Packager changes:

- Split the bindings into separate file for easier support and to keep the driver program size down.
- Conditional include of bindings depending on if  --nobindings is specified.
- Packager: AOT ninja build modifications.
   - Add corebindings build dependency
   - Add bindings_support.j build dependency
   - Remove dependency to add Bindings initialisation to build and mono_config.js
 - Modify runtime.js and runtime-tests.js to remove the bindings initialisation code.

### CODE REVIEW

File: sdks/wasm/framework/src/WebAssembly.Bindings/Core/Float32Array.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+		public Float32Array (ArrayBuffer buffer, int byteOffset) : base (buffer, byteOffset) { }
+
+		public Float32Array (ArrayBuffer buffer, int byteOffset, int length) : base (buffer, byteOffset, length) { }
+
+		internal Float32Array (IntPtr js_handle) : base (js_handle) { }
+
+
+		// Define the indexer to allow client code to use [] notation.
+		public float? this [int i] {
```
Comment (kg) [inline sdks/wasm/framework/src/WebAssembly.Bindings/Core/Float32Array.cs]: do we need to duplicate all the array logic like this? i'm guessing there's some reason why we can't use generics

Comment (kjpou1) [reply sdks/wasm/framework/src/WebAssembly.Bindings/Core/Float32Array.cs]: The problem was the conversion.  The return values are int or double and were causing a problem with a way to unbox them.

Comment (kg) [reply sdks/wasm/framework/src/WebAssembly.Bindings/Core/Float32Array.cs]: I see, makes sense. Can you hoist all the logic except the cast down into a function in the base class? Not a requirement to merge, but would cut out a lot of duplication.

Comment (kjpou1) [reply sdks/wasm/framework/src/WebAssembly.Bindings/Core/Float32Array.cs]: Yes, I agree.  I was planning on re-visiting that part as well.  Just wanted to get this out here for review.

Comment (kg) [reply sdks/wasm/framework/src/WebAssembly.Bindings/Core/Float32Array.cs]: Sure thing. Not something I consider a merge-blocker, anyway. Big PR so I haven't looked at the whole thing yet but it looks good so far.

Comment (kjpou1) [reply sdks/wasm/framework/src/WebAssembly.Bindings/Core/Float32Array.cs]: Array logic moved.

**Comment by Suchiman** [comment]
@kjpou1 
> Typed Array memory can not be directly accessed but instead needs to be transferred to and from emscripten's heap.

Why? In Blazor i've used https://github.com/aspnet/Blazor/blob/5deb85df8784800742e7a6d68575c67dd6fedb5a/src/Microsoft.AspNetCore.Blazor.Browser.JS/src/Platform/Mono/MonoPlatform.ts#L96-L100
which allowed to access the managed array from JS directly (similiar to how Span works).
Your `typedarray_copy_to` does this as well (to copy from a non emscripten heap to the emscripten heap / ArrayBuffer).

**Comment by kg** [comment]
@Suchiman that won't work for foreign typed arrays and array-like buffers, i.e. the ones used in Web Audio and HTML5 Canvas

**Comment by kjpou1** [comment]
@Suchiman 

One example of accessing foreign Typed Arrays are when accessing `WebSocket` interface `onmessage` event for binaryType blob.

The [FileReader.readAsArrayBuffer()
](https://developer.mozilla.org/en-US/docs/Web/API/FileReader/readAsArrayBuffer) reads the `ArrayBuffer` from the `Blob` of the event.  To access the array buffer data it needs to be copied to the heap to be referenced.  https://github.com/WebAssembly/design/issues/1162

What this PR attempts to accomplish is make this as seamless as possible by adding methods that allow for brokering this information in and out of heap.  Allowing for CopyTo, Slice, CopyFrom, ToArray etc.. As well as ArraySegment and Span.

The first implementation hid the details of this marshalling at the sacrifice of marshalling a full copy of  the typed array whether or not it was needed.  This lead to in some cases a lot of memory being allocated depending on the interface.  

Hope this helps

**Comment by kjpou1** [comment]
@monojenkins build failed

**Comment by kjpou1** [comment]
@monojenkins build failed

File: sdks/wasm/framework/src/WebAssembly.Bindings/Core/Int8Array.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+		internal Int8Array (IntPtr js_handle) : base (js_handle)
+		{ }
+
+		/// <summary>
+		/// Defines an implicit conversion of a <see cref="ArraySegment{T}"/> to a <see cref="T:WebAssembly.Core.Int8Array"/>/>
+		/// </summary>
+		/// <returns>The implicit.</returns>
+		/// <param name="segment">ArraySegment</param>
+		public static implicit operator Int8Array (ArraySegment<sbyte> segment) => From (segment);
```
[the remark this answers is not in the document]
Comment (kjpou1) [reply sdks/wasm/framework/src/WebAssembly.Bindings/Core/Int8Array.cs]: No, thanks for the catch.  Missed a cleanup.

File: sdks/wasm/framework/src/WebAssembly.Bindings/Core/TypedArray.cs [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+					throw new JSException ((string)res);
+				return (T)res;
+			}
+
+		}
+
+		/// <summary>
+		/// Copies to.
+		/// </summary>
+		/// <returns>The to.</returns>
```
[the remark this answers is not in the document]
Comment (kjpou1) [reply sdks/wasm/framework/src/WebAssembly.Bindings/Core/TypedArray.cs]: addressed in recent commit

File: sdks/wasm/framework/src/WebAssembly.Bindings/Core/TypedArray.cs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+			if (exception != 0)
+				throw new JSException ((string)res);
+			return (int)res / Marshal.SizeOf<U> ();
+
+		}
+
+		/// <summary>
+		/// Copies from.
+		/// </summary>
+		/// <returns>The from.</returns>
```
[the remark this answers is not in the document]
Comment (kjpou1) [reply sdks/wasm/framework/src/WebAssembly.Bindings/Core/TypedArray.cs]: addressed in recent commit

File: sdks/wasm/framework/src/WebAssembly.Bindings/Core/TypedArray.cs [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+			} 
+			else
+				return null;
+		}
+
+		/// <summary>
+		/// To a span.
+		/// </summary>
+		/// <returns>The span.</returns>
+		public Span<U> ToSpan ()
```
[the remark this answers is not in the document]
Comment (kjpou1) [reply sdks/wasm/framework/src/WebAssembly.Bindings/Core/TypedArray.cs]: Addressed in recent commit

File: sdks/wasm/framework/src/WebAssembly.Bindings/Core/Int8Array.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kjpou1) [reply sdks/wasm/framework/src/WebAssembly.Bindings/Core/Int8Array.cs]: Addressed in recent commit

File: sdks/wasm/framework/src/WebAssembly.Bindings/Core/TypedArray.cs [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+			return (int)res / Marshal.SizeOf<U> ();
+
+		}
+
+		/// <summary>
+		/// Copies from.
+		/// </summary>
+		/// <returns>The from.</returns>
+		/// <param name="span">Span.</param>
+		public unsafe int CopyFrom (Span<U> span)
```
[the remark this answers is not in the document]
Comment (kjpou1) [reply sdks/wasm/framework/src/WebAssembly.Bindings/Core/TypedArray.cs]: Addressed in recent commit

File: sdks/wasm/framework/src/WebAssembly.Bindings/Runtime.cs [hunk 7]
```diff
@@ -76,23 +117,56 @@ public static JSObject NewJSObject (JSObject js_func_ptr = null, params object [
...
+				bound_objects [js_id] = obj = (JSObject)jsobjectnew.Invoke (new object [] { (IntPtr)js_id });
+			}
+			return (int)(IntPtr)obj.Handle;
+		}
+
 		static int UnBindJSObject (int js_id)
 		{
-			if (bound_objects.TryGetValue (js_id, out var obj)) {
+			if (bound_objects.ContainsKey (js_id)) {
+				var obj = bound_objects [js_id];
```
Comment (Suchiman) [inline sdks/wasm/framework/src/WebAssembly.Bindings/Runtime.cs]: Why was this changed to the less efficient `ContainsKey(…) + this[…]` pattern?

Comment (kjpou1) [reply sdks/wasm/framework/src/WebAssembly.Bindings/Runtime.cs]: My bad after conflict resolution.  Thanks for the catch.

Comment (kjpou1) [reply sdks/wasm/framework/src/WebAssembly.Bindings/Runtime.cs]: Should be fixed

**Comment by kjpou1** [comment]
@monojenkins build failed

File: sdks/wasm/framework/src/WebAssembly.Bindings/Core/ArrayBuffer.cs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+		/// <summary>
+		/// Initializes a new instance of the <see cref="T:WebAssembly.Core.ArrayBuffer"/> class.
+		/// </summary>
+		/// <param name="js_handle">Js handle.</param>
+		internal ArrayBuffer (IntPtr js_handle) : base (js_handle)
+		{ }
+
+		/// <summary>
+		/// Gets the length of the byte.
```
Comment (lewing) [inline sdks/wasm/framework/src/WebAssembly.Bindings/Core/ArrayBuffer.cs]:
```suggestion
		/// The length of an ArrayBuffer in bytes. 
```

File: sdks/wasm/framework/src/WebAssembly.Bindings/Core/ArrayBuffer.cs [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+		/// </summary>
+		/// <value>The length of the byte.</value>
```
Comment (lewing) [inline sdks/wasm/framework/src/WebAssembly.Bindings/Core/ArrayBuffer.cs]:
```suggestion
		/// <value>The length of the underlying ArrayBuffer in bytes.</value>
```

File: sdks/wasm/framework/src/WebAssembly.Bindings/Core/Array.cs [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+		/// <returns>The push.</returns>
+		/// <param name="elements">Elements.</param>
+		public int Push (params object[] elements) => (int)Invoke ("push", elements);
+		/// <summary>
+		/// Pop this instance.
+		/// </summary>
+		/// <returns>The pop.</returns>
+		public int Pop () => (int)Invoke ("pop");
+		/// <summary>
+		/// Shift this instance.
```
Comment (lewing) [inline sdks/wasm/framework/src/WebAssembly.Bindings/Core/Array.cs]:
```suggestion
		/// Remove the first element of the Array and return that element
```

File: sdks/wasm/framework/src/WebAssembly.Bindings/Core/Array.cs [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+		/// </summary>
+		/// <returns>The shift.</returns>
```
Comment (lewing) [inline sdks/wasm/framework/src/WebAssembly.Bindings/Core/Array.cs]:
```suggestion
		/// <returns>The removed element</returns>
```

File: sdks/wasm/framework/src/WebAssembly.Bindings/Host/HostObjectInterface.cs [hunk 12]
```diff
@@ -0,0 +1,6 @@
+﻿using System;
+namespace WebAssembly.Host {
+	public interface HostObjectInterface {
```
Comment (lewing) [inline sdks/wasm/framework/src/WebAssembly.Bindings/Host/HostObjectInterface.cs]: Why not IHostObject?

Comment (kjpou1) [reply sdks/wasm/framework/src/WebAssembly.Bindings/Host/HostObjectInterface.cs]: Absolutely right.

Comment (kjpou1) [reply sdks/wasm/framework/src/WebAssembly.Bindings/Host/HostObjectInterface.cs]: Addressed in commit

**Comment by lewing** [comment]
@monojenkins build failed
